### PR TITLE
Add adjustable delay between market actions

### PR DESCRIPTION
**Pull request by Sadzurami**
Steam changed rate limits on various endpoints, including market

Would be cool to see adjustable delay between market actions

ref: https://github.com/Nuklon/Steam-Economy-Enhancer/issues/219

### CODE REVIEW

**Comment by Sadzurami** [comment]
To be honest, i'm not sure is it worth to apply this delays for price loading requests, or not

**Comment by Sadzurami** [comment]
I removed delays for price loading for now

If someone will see limits for that, will be easy to apply delays back

File: code.user.js [hunk 1]
```diff
@@ -1250,7 +1252,11 @@
                                 css('background', COLOR_ERROR);
                         }
 
-                        next();
+                        const delay = sellQueue.length() > 0 
+                            ? parseInt(getSettingWithDefault(SETTING_DELAY_BETWEEN_MARKET_ACTIONS), 10) * 1000
+                            : 0;
+
+                        setTimeout(() => next(), delay);
```
Comment (xPaw) [inline code.user.js:1259]: Verify that delay is not `isNaN`

**Comment by Sadzurami** [comment]
> Do you know which calls trigger the rate limit (is it 429 response)? If so, would be good to detect specifically that and increase the delay greatly and/or alert the user somehow.

`https://steamcommunity.com/market/removelisting/:id`
`https://steamcommunity.com/market/sellitem/`

429 http code

maybe even (not sure)
`https://steamcommunity.com/market/market/cancelbuyorder/`

but i don't think increasing delays after limits is good idea, because we don't know when will be next limits window

**Comment by xPaw** [comment]
If a 429 is hit, there's no point even trying to send a request (especially with a user configured delay), you can force next request to be delayed by like 2 minutes or whatever the rate limit is.

**Comment by Sadzurami** [comment]
> If a 429 is hit, there's no point even trying to send a request (especially with a user configured delay), you can force next request to be delayed by like 2 minutes or whatever the rate limit is.

made some changes
please, check it, if you have some time

File: code.user.js [hunk 2]
```diff
@@ -488,6 +498,13 @@
                 error: function() {
                     return callback(ERROR_FAILED);
                 },
+                statusCode: {
```
Comment (xPaw) [inline code.user.js:501]: Seems weird to list all the status codes, I'd check for 429 in the error callback instead. (I think it goes into error callback?)

Comment (Sadzurami) [reply code.user.js]: made some changes, now statuscode checks via xhr.status

**Comment by Sadzurami** [comment]
It seems that Steam is throttling other /market/ requests (I checked myself). So I return delays back to price and listings loading requests.

File: code.user.js [hunk 3]
```diff
@@ -455,15 +458,18 @@
...
                     callback(ERROR_FAILED, data);
                 } else {
                     callback(ERROR_SUCCESS, data);
                 }
             },
-            error: function(data) {
-                return callback(ERROR_FAILED, data);
+            error: function(xhr) {
+                marketRateLimitReached = xhr.status === 429;
+                return callback(ERROR_FAILED, xhr);
```
Comment (xPaw) [inline code.user.js:472]: Passing xhr seems wrong (you didn't change it I know), probably meant to be passing the response text or something? Could you verify what the callback expects on error please.

Comment (Sadzurami) [reply code.user.js]: > Passing xhr seems wrong (you didn't change it I know), probably meant to be passing the response text or something? Could you verify what the callback expects on error please.

yes, ig it's not correct
i changed this behaviour on last commits

**Comment by Nuklon** [comment]
Thanks for this PR, I think a fix for throttling is OK, but I don't see any point in adding a configurable setting.
It should:
- Use a timeout that doesn't cause throttling.
- Delay requests after being throttled (similar to what @xPaw said).

**Comment by Sadzurami** [comment]
> Thanks for this PR, I think a fix for throttling is OK, but I don't see any point in adding a configurable setting. It should:
> 
> * Use a timeout that doesn't cause throttling.
> * Delay requests after being throttled (similar to what @xPaw said).

Agree with you.
The main purpose of having configurable setting is https://github.com/Nuklon/Steam-Economy-Enhancer/issues/219.

Also, we don't know exactly steam limits. I guess 1 second between requests will be enough, and i tested 1s and didn't see rate limits, but it's not guarantee.

And, there is possibility that steam will change limits in future, so, having configurable settings is not that bad. But, yea, for this case better to have setting not to only market request, but for all requests (now 300ms by default).

**Comment by Nuklon** [comment]
If Steam changes it again we can release a new version to address it, I don't think this needs to be configurable as it should work out of the box.

**Comment by Sadzurami** [comment]
> If Steam changes it again we can release a new version to address it, I don't think this needs to be configurable as it should work out of the box.

Alr, i'll make commit with changes.

**Comment by Sadzurami** [comment]
@Nuklon looks like it's done

File: code.user.js [hunk 4]
```diff
@@ -109,6 +109,63 @@
         }
     }
 
+    function request(url, options, callback) {
+        let delayBetweenRequests = 300;
+        let requestStorageHash = 'request:last';
+
+        if (url.startsWith('https://steamcommunity.com/market/')) {
+            requestStorageHash = 'request:last:steamcommunity.com/market';
```
Comment (xPaw) [inline code.user.js:117]: session storage is probably enough here, and I'd prefix the strings with "see" or something.

Comment (Sadzurami) [reply code.user.js]: done

**Comment by Sadzurami** [comment]
returned back localStorage to request function

it will help prevent limits, if two tabs are open in same time, for example market and inventory

**Comment by xPaw** [comment]
That's my bad, I thought sessionStorage was shared between tabs.

**Comment by Sadzurami** [comment]
@Nuklon please, take a look to this pr request

File: code.user.js [hunk 5]
```diff
@@ -109,6 +109,63 @@
...
     }
 
+    function request(url, options, callback) {
+        let delayBetweenRequests = 300;
+        let requestStorageHash = 'see:request:last';
+
+        if (url.startsWith('https://steamcommunity.com/market/')) {
+            requestStorageHash = `${requestStorageHash}:steamcommunity.com/market`;
+            delayBetweenRequests = 1000;
+        };
```
Comment (Nuklon) [inline code.user.js:119]: Please remove all `;` after if statements.

**Pull request merged**